eth_worker/eth_manager/blockchain_sync/blockchain_sync.py

```diff
@@ -0,0 +1,134 @@
+        number_of_chunks = ceil(number_of_blocks_to_get/sync_const.BLOCKS_PER_REQUEST)
+
+        for chunk in range(number_of_chunks):
+            floor = max_fetched_block + (chunk * sync_const.BLOCKS_PER_REQUEST) + 1
+            ceiling = max_fetched_block + ((chunk + 1) * sync_const.BLOCKS_PER_REQUEST)
+            if ceiling > latest_block:
+                ceiling = latest_block
+            process_chunk(f, floor, ceiling)
+            persistence_module.set_filter_max_block(f.id, ceiling)
+    return True
+
+# Gets history for given range, and runs handle_transaction on all of them
+# This is the second stage in the third party transaction processing pipeline!
+def process_chunk(filter, floor, ceiling):
+    transaction_history = get_blockchain_transaction_history(
+            filter.contract_address, 
+            floor, 
+            ceiling, 
+            filter.filter_parameters,
+            filter.id
+        )
+    for transaction in transaction_history:
+        handle_transaction(transaction, filter)
+
+# Processes newly found transaction event
+# Creates database object for transaction
+# Calls webhook
+# Sets sync status (whether or not webhook was successful)
+# Fallback if something goes wrong at this level: `is_synchronized_with_app` flag. Can batch unsynced stuff
+def handle_transaction(transaction, filter):
+    # Check if transaction already exists (I.e. already synchronized, or first party transactions)
+    transaction_object = persistence_module.get_transaction(hash=transaction.transactionHash.hex())
+    if transaction_object and transaction_object.is_synchronized_with_app:
+        return True
+    transaction_object = persistence_module.create_external_transaction(
+        status = 'SUCCESS',
+        block = transaction.blockNumber,
+        hash = str(transaction.transactionHash.hex()),
+        contract_address = transaction.address,
+        is_synchronized_with_app = False,
+        recipient_address = transaction.args['to'],
+        sender_address = transaction.args['from'],
+        amount = float(transaction.args['value'])/(10**filter.decimals)
+    )
+    webook_resp = call_webhook(transaction_object)
+    # Transactions which we fetched, but couldn't sync for whatever reason won't be marked as completed
+    # in order to be retryable later
+    if webook_resp.ok:
+        persistence_module.mark_transaction_as_completed(transaction_object)
+    return True
+
+# Gets blockchain transaction history for given range
+# Fallback if something goes wrong at this level: block-tracking table
+def get_blockchain_transaction_history(contract_address, start_block, end_block = 'lastest', argument_filters = None, filter_id = None):
+    # Creates DB objects for every block to monitor status
+    config.logg.info(f'Fetching block range {start_block} to {end_block} for contract {contract_address}')
+
+    persistence_module.add_block_range(start_block, end_block, filter_id)
```

Comment:
With overlapping calls (from above), it is possible to add multiple identical blocks.

Reply:
Covered by adding a lock